core/src/main/scala/org/ensime/core/AnalyzerManager.scala

```diff
@@ -0,0 +1,339 @@
+        historyOfModule += module -> LoadedFilesData(Map.empty, Loaded)
+      }
+    case req @ UnloadAllReq =>
+      analyzers.foreach {
+        case (_, analyzer) => analyzer forward req
+      }
+    case req @ TypecheckModule(moduleId) =>
+      withExistingModule(moduleId, req)((req, module) => {
+        getOrSpawnNew(module) forward req
+        historyOfModule += module -> LoadedFilesData(Map.empty, Loaded)
+      })
+    case req @ UnloadModuleReq(moduleId) =>
+      withExistingModule(moduleId, req)((req, module) => {
+        getOrSpawnNew(module) forward req
+        val previousData = historyOfModule(module)
+        historyOfModule += module -> update(previousData) { (statusOfFile, default) =>
+          default match {
+            case NotLoaded =>
+              LoadedFilesData(statusOfFile.map(m => m._1 -> m._2.append(Removed)), default)
+            case _ =>
+              LoadedFilesData(Map.empty, Removed)
+          }
+        }
+      })
+    case req @ RemoveFileReq(file: File) =>
+      val f = SourceFileInfo(RawFile(file.toPath), None, None)
+      withExistingModuleFor(f, req)((req, module) => {
+        getOrSpawnNew(module) forward req
+        val previousData = historyOfModule(module)
+        historyOfModule += module -> update(previousData) { (statusOfFile, default) =>
+          LoadedFilesData(statusOfFile + (f -> statusOfFile.getOrElse(f, default).append(Removed)), default)
+        }
+      })
+    case req @ TypecheckFileReq(fileInfo) =>
+      withExistingModuleFor(fileInfo, req)((req, module) => {
+        getOrSpawnNew(module) forward req
+        val previousData = historyOfModule(module)
+        historyOfModule += module -> update(previousData) { (statusOfFile, default) =>
+          LoadedFilesData(statusOfFile + (fileInfo -> statusOfFile.getOrElse(fileInfo, default).append(Loaded)), default)
+        }
+      })
+    case req @ TypecheckFilesReq(files) =>
+      var missingSource: Boolean = files.exists(getModule(_).isEmpty)
+      if (missingSource)
+        sender ! EnsimeServerError("Update .ensime file.")
+      else {
+        val filesPerModule: Map[EnsimeProject, List[Either[File, SourceFileInfo]]] = files.groupBy(getModule).map(x => x._1.get -> x._2)
+        for ((module, list) <- filesPerModule) {
+          val analyzer = getOrSpawnNew(module)
+          analyzer ! TypecheckFilesReq(list)
+          if (!list.map(toSourceFileInfo).exists(!_.file.exists)) {
+            val previousData = historyOfModule(module)
+            historyOfModule += module -> update(previousData) { (statusOfFile, default) =>
+              val addToStatusOfFile = list.map(toSourceFileInfo).map(f => f -> statusOfFile.getOrElse(f, default).append(Loaded))
+              LoadedFilesData(statusOfFile ++ addToStatusOfFile, default)
+            }
+          }
+
+        }
+        context.become(collector[List[String]](filesPerModule.size, Nil, sender)((newResponse, aggregate) => {
```

Comment:
a concern I have about all these `become` calls is that it means ensime becomes effectively single threaded... the analyzer manager holds back all calls until this one responds, instead of being able to process many at a time. We'll need to think of another way of doing the results collecting.